### PR TITLE
Post-training Quantization

### PR DESCRIPTION
**Pull request by sydarb**
Added the codes for post-training quantization based on block reconstruction method.
 
Changes made outside quantize folder:
- utils/__init__.py : added seed for multi-gpu
- datasets/classfication/imagenet.py : resolved ImportError
- datasets/classfication/__init__.py :       ""
- models/mobilenetv2.py : added MobileNetV2 model

Files added inside quantize folder:
- quantize.py : base class for quantization along with functions that can be used in other quantization methods.
- brecq.py : base class for block reconstruction method.
- quant_model.py : wrapper classes for building quantization modules and parameters. changes can be made to add support for more models. could be used by other quant methods.
- data_hooks.py : hook manager to store full precision outputs and gradients from quantized model. can be used by other quant methods.
- reconstruct.py : finetunes quantized weights layer and block wise. 

All features tested successfully on Imagenet dataset with ResNet50 model. Test notebook can be found inside '/experiments/brecq' folder.

### CODE REVIEW

File: trailmet/algorithms/quantize/quantize.py [hunk 1]
```diff
@@ -0,0 +1,181 @@
...
+
+    def is_bn(self, m):
+        return isinstance(m, nn.BatchNorm2d) or isinstance(m, nn.BatchNorm1d)
+
+
+    def is_absorbing(self, m):
+        return (isinstance(m, nn.Conv2d)) or isinstance(m, nn.Linear)
+
+
+    def search_fold_and_remove_bn(self, model: nn.Module):
```
Comment (rishabh-16) [inline trailmet/algorithms/quantize/quantize.py:92]: add docstring

File: trailmet/algorithms/quantize/quantize.py [hunk 2]
```diff
@@ -0,0 +1,181 @@
...
+                # set the bn module to straight through
+                setattr(model, n, StraightThrough())
+            elif self.is_absorbing(m):
+                prev = m
+            else:
+                prev = self.search_fold_and_remove_bn(m)
+        return prev
+
+
+    def search_fold_and_reset_bn(self, model: nn.Module):
```
Comment (rishabh-16) [inline trailmet/algorithms/quantize/quantize.py]: add docstring

File: trailmet/algorithms/quantize/quantize.py [hunk 3]
```diff
@@ -0,0 +1,181 @@
...
+# the lp_loss, accuracy and test functions can be shifted to the BaseAlogorithm class in ./algorithms.py
+    
+    def lp_loss(pred, tgt, p=2.0, reduction='none'):
+        """loss function measured in Lp Norm"""
+        if reduction == 'none':
+            return (pred-tgt).abs().pow(p).sum(1).mean()
+        else:
+            return (pred-tgt).abs().pow(p).mean()
+
+    def accuracy(self, output, target, topk=(1,)):
```
Comment (rishabh-16) [inline trailmet/algorithms/quantize/quantize.py]: its a general function, can we move it to algorithms or utils?

File: trailmet/algorithms/quantize/quantize.py [hunk 4]
```diff
@@ -0,0 +1,181 @@
...
+            pred = pred.t()
+            correct = pred.eq(target.view(1, -1).expand_as(pred))
+
+            res = []
+            for k in topk:
+                correct_k = correct[:k].reshape(-1).float().sum(0, keepdim=True)
+                res.append(correct_k.mul_(100.0 / batch_size))
+            return res
+
+    def test(self, model, dataloader, loss_fn=None, device=None):
```
Comment (rishabh-16) [inline trailmet/algorithms/quantize/quantize.py]: this is already implemented, can we combine the both?

**Review by rishabh-16 — CHANGES_REQUESTED** [review]

File: trailmet/algorithms/quantize/reconstruct.py [hunk 5]
```diff
@@ -0,0 +1,355 @@
...
+
+    if not act_quant:
+        # Replace weight quantizer to AdaRoundQuantizer
+        layer.weight_quantizer = AdaRoundQuantizer(uaq=layer.weight_quantizer, round_mode=round_mode,
+                                                weight_tensor=layer.org_weight.data)
+        layer.weight_quantizer.soft_targets = True
+
+        # Set up optimizer
+        opt_params = [layer.weight_quantizer.alpha]
+        optimizer = torch.optim.Adam(opt_params)
```
Comment (rishabh-16) [inline trailmet/algorithms/quantize/reconstruct.py]: consider adding optimizer to kwargs

File: trailmet/algorithms/quantize/reconstruct.py [hunk 6]
```diff
@@ -0,0 +1,355 @@
...
+                module.weight_quantizer = AdaRoundQuantizer(uaq=module.weight_quantizer, round_mode=round_mode,
+                                                            weight_tensor=module.org_weight.data)
+                module.weight_quantizer.soft_targets = True
+
+        # Set up optimizer
+        opt_params = []
+        for name, module in block.named_modules():
+            if isinstance(module, QuantModule):
+                opt_params += [module.weight_quantizer.alpha]
+        optimizer = torch.optim.Adam(opt_params)
```
Comment (rishabh-16) [inline trailmet/algorithms/quantize/reconstruct.py]: consider adding optimizer to kwargs

**Review by rishabh-16 — CHANGES_REQUESTED** [review]

File: trailmet/algorithms/quantize/quantize.py (hunk 1 above)
Comment (sydarb) [reply trailmet/algorithms/quantize/quantize.py:92]: done in e8b1f0bec3d8d137435419fb9a5fb054ab203cb3

File: trailmet/algorithms/quantize/quantize.py (hunk 2 above)
Comment (sydarb) [reply trailmet/algorithms/quantize/quantize.py]: done in e8b1f0bec3d8d137435419fb9a5fb054ab203cb3

Comment (sydarb) [reply trailmet/algorithms/quantize/quantize.py]: method was obsolete so removed it

File: trailmet/algorithms/quantize/reconstruct.py (hunk 5 above)
Comment (sydarb) [reply trailmet/algorithms/quantize/reconstruct.py]: done in 33615aff540d4b3883842eca67b1a56538f591a1

File: trailmet/algorithms/quantize/reconstruct.py (hunk 6 above)
Comment (sydarb) [reply trailmet/algorithms/quantize/reconstruct.py]: done in 33615aff540d4b3883842eca67b1a56538f591a1

File: trailmet/algorithms/quantize/quantize.py (hunk 3 above)
Comment (sydarb) [reply trailmet/algorithms/quantize/quantize.py]: lp_loss is also a generic function, should it be moved to algorithms along with test and accuracy?

https://github.com/heisenbug237/trailmet/blob/33615aff540d4b3883842eca67b1a56538f591a1/trailmet/algorithms/quantize/quantize.py#L36

Comment (rishabh-16) [reply trailmet/algorithms/quantize/quantize.py]: yes please shift it

Comment (sydarb) [reply trailmet/algorithms/quantize/quantize.py]: done in e52f86293114e05d095189e26f3c0dcce71d6235

File: trailmet/algorithms/quantize/quantize.py (hunk 4 above)
Comment (sydarb) [reply trailmet/algorithms/quantize/quantize.py]: done in e52f86293114e05d095189e26f3c0dcce71d6235

**Comment by Arnav0400** [comment]
@heisenbug237 is this PR still valid?
